File: Ex04/ex04.c
/*
** PERSO, 2023
** emmackathon
** File description:
** ec06
*/

#include <stdlib.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

typedef struct vaches_s {
    int age;
    int masse;
    char *name;
    int note;
    int count;
} vaches_t;

vaches_t **read_all()
{
   char *line = NULL;
   size_t size;
   int count = -2;
   vaches_t **vaches = NULL;
  
	while (getline(&line, &size, stdin) != -1) {
        count += 1;
        if (count == -1)
            vaches = malloc(sizeof(vaches_t *) * atoi(line));
        else {
            vaches[count] = malloc(sizeof(vaches_t));
            vaches[count]->name = strdup(strtok(line, " "));
            vaches[count]->age = atoi(strtok(NULL, " "));
            vaches[count]->masse = atoi(strtok(NULL, " "));
            vaches[count]->note = atoi(strtok(NULL, " "));
            vaches[count]->note += atoi(strtok(NULL, " "));
        }
	}
	vaches[0]->count = count;
    return vaches;
}

char *check_vache(vaches_t **vaches, int count)
{
    int best = 0;
    char *best_name = NULL;
 
    for (int i = 0; i < count + 1; i++) {
        if (vaches[i]->masse > 1500 || vaches[i]->masse < 1250)
            continue;
        if (vaches[i]->age < 2 || vaches[i]->age > 5)
            continue;
        if (vaches[i]->note > best) {
            best_name = strdup(vaches[i]->name);
            best = vaches[i]->note;
        } else if (vaches[i]->note == best) {
            best_name = strcat(best_name, " ");
            best_name = strcat(best_name, strdup(vaches[i]->name));
        }
    }
    return best_name;
}

int main() {
    vaches_t **vaches = read_all();
    int count = vaches[0]->count;
    char *best_name = check_vache(vaches, count);

    printf("%s\n", best_name);
    
    return 0;
}

File: Ex04/ex04.py
#!/usr/bin/env python3
import sys


def read_all() -> list[list[str]]:
    lines: list[str] = []
    all_lines: list[list[str]] = []
    for line in sys.stdin:
        lines = line.rstrip("\n").split(" ")
        if len(lines) != 5:
            continue
        age = int(lines[1])
        masse = int(lines[2])
        if age > 5 or age < 2 or masse > 1500 or masse < 1250:
            continue
        note1 = int(lines[3])
        note2 = int(lines[4])
        if note1 < 0 or note2 < 0 or note1 > 20 or note2 > 20:
            continue
        all_lines.append(lines)
    return all_lines


def sort_cow(all_lines: list[list[str]]) -> list[str]:
    score: int = 0
    for i in all_lines:
        new = int(i[3]) + int(i[4])
        if new > score:
            best = []
            best.append(i[0])
            score = new
        elif new == score:
            best.append(i[0])
    return best


def print_best(best: list[str]) -> int:
    print(best[0], end="")
    for i in best[1::]:
        print(f" {i}", end="")
    print()
    return 0


def main() -> int:
    lines: list[list[str]] = read_all()
    best: list[str] = sort_cow(lines)
    print_best(best)
    return 0


main()
